server/libs/domain/src/search/search.repository.ts

```diff
@@ -0,0 +1,33 @@
+import { AlbumEntity, AssetEntity } from '@app/infra/db/entities';
+
+export enum SearchCollection {
+  ASSETS = 'assets',
+  ALBUMS = 'albums',
+}
+
+export interface SearchFilter {
+  userId: string;
+}
+
+export interface SearchResult<T> {
+  page: number;
+  items: T[];
+}
+
+export type SearchCollectionIndexStatus = Record<SearchCollection, boolean>;
+
+export const ISearchRepository = 'ISearchRepository';
+
+export interface ISearchRepository {
+  setup(): Promise<void>;
+  checkMigrationStatus(): Promise<SearchCollectionIndexStatus>;
+
+  index(collection: SearchCollection.ASSETS, item: AssetEntity): Promise<void>;
+  index(collection: SearchCollection.ALBUMS, item: AlbumEntity): Promise<void>;
```

Comment:
Could the SearchCollection<->Entity relation somehow be included in the enum itself? I imagine with generics probably, but those are a bit too over my head to actually offer an approach 😅

Reply:
Possibly, yes.

Reply:
The main issue is that genetics is all stripped out at compile time, so you would have to infer the actual entity type some other way (for use in the implementation)

Reply:
Dang typescript strikes again with it's compiletime-only types >.< Gets me every time lol